### PR TITLE
Added support for MySQL Information Schema

### PR DESCRIPTION
**Pull request by Dharmil-searce**
Imported [GO-MySQL-Driver](https://github.com/go-sql-driver/mysql) to use [database/sql](https://golang.org/pkg/database/sql/) package.

### CODE REVIEW

File: mysql/infoschema.go [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+}
+
+// ProcessSQLData performs data conversion for source database
+// 'db'. For each table, we extract data using a "SELECT (colNamesList)" query,
+// convert the data to Spanner data (based on the source and Spanner
+// schemas), and write it to Spanner.  If we can't get/process data
+// for a table, we skip that table and process the remaining tables.
+//
+// Note that the database/sql library has a somewhat complex model for
+// returning data from rows.Scan. We pass *sql.RawBytes to rows.scan.
```
Comment (nevinheintze) [inline mysql/infoschema.go:56]: Suggest dropping "Note that the database/sql library has a somewhat complex model for
returning data from rows.Scan." 

That comment was from the postgres code where we had to deal with different encodings for scalars and arrays.

File: mysql/infoschema.go [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+	_ "github.com/go-sql-driver/mysql" // The driver should be used via the database/sql package.
+	_ "github.com/lib/pq"
+)
+
+// ProcessInfoSchema performs schema conversion for source database
+// 'db'. We assume that the source database supports information
+// schema tables. These tables are a broadly supported ANSI standard,
+// and we use them to obtain source database's schema information.
+func ProcessInfoSchema(conv *internal.Conv, db *sql.DB) error {
+	// TODO: refactor to use the set of tables computed by
```
Comment (nevinheintze) [inline mysql/infoschema.go:33]: Cut-and-paste error?
This comment doesn't make sense here (there is already a copy of this comment in ProcessSQLData, where it belongs).

File: mysql/infoschema.go [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+	"database/sql"
+	"fmt"
+	"strings"
+
+	"github.com/cloudspannerecosystem/harbourbridge/internal"
+	"github.com/cloudspannerecosystem/harbourbridge/schema"
+	_ "github.com/go-sql-driver/mysql" // The driver should be used via the database/sql package.
+	_ "github.com/lib/pq"
+)
+
```
Comment (nevinheintze) [inline mysql/infoschema.go:27]: There is significant overlap between mysql/infoschema.go and postgres/inforschema.go.

Can you summarize the differences?

Probably worth a TODO to capture this.

Comment (Dharmil-searce) [reply mysql/infoschema.go:27]: Main differences between postgres/infoschema.go and mysql/infoschema.go:

- SQL statements.
   - Difference in sql statements `ProcessSQLData` and `getColumns`.
   - In `SetRowStats` -  dropped the `" "` from table_schema and table_name as per mysql syntax.

- Different way to read data i.e. instead of interface{} we are using sql.RawBytes{}.

- Different ways of handling data conversion. Instead of `ConvertSQLRow`,`cvtSQLScalar` and `cvtSQLArray` function used for data conversion in postgres, mysql is converting row values to string and then using `ProcessDataRow` function of `mysql/data.go` file for data conversion. Hence, removed the redundancy of data conversion functions by directly using functions in `mysql/data.go`.

- Added a function `buildColNameList` to handle spatial datatypes. 

- Different list of system tables to be ignored in `getTables` function. 
   - Mysql - `{"information_schema", "mysql", "performance_schema", "sys"}` 
   - Postgres - `{"information_schema", "postgres", "pg_catalog", "pg_temp_1", "pg_toast", "pg_toast_temp_1"}`

- Do not need to drop `public` prefix in table name because mysql do not support `public`.

- `valsToStrings` is different because postgres uses `[]interface{}` argument, while mysql uses `[]sql.RawBytes` argument.

File: mysql/infoschema.go (hunk 2 above)
Comment (Dharmil-searce) [reply mysql/infoschema.go]: My Bad. Removed it.

File: mysql/infoschema.go (hunk 1 above)
Comment (Dharmil-searce) [reply mysql/infoschema.go]: Done. Formatted the comment.

File: mysql/toddl.go [hunk 4]
```diff
@@ -136,10 +136,10 @@ func toSpannerType(conv *internal.Conv, id string, mods []int64) (ddl.ScalarType
 		}
 		return ddl.String{Len: ddl.MaxLength{}}, nil
 	case "text", "tinytext", "mediumtext", "longtext":
-		maxExpectedMods(0)
+		maxExpectedMods(1)
```
Comment (nevinheintze) [inline mysql/toddl.go:139]: Can you give some background on this change: can text types have a modifier?
Can you give an example?

Comment (Dharmil-searce) [reply mysql/toddl.go:139]: Though we would not be specifying mods for `text` type, but MySQL by default will put max value for `text` type in `CHARACTER_MAXIMUM_LENGTH` column of `information_schema.COLUMNS` table. Eg: 
```
mysql> select c.column_name,c.data_type,c.is_nullable,c.column_default, c.character_maximum_length, c.numeric_precision, c.numeric_scale from information_schema.COLUMNS c where table_schema='testdb' and COLUMN_NAME LIKE '%text';
+-------------+------------+-------------+----------------+--------------------------+-------------------+---------------+
| COLUMN_NAME | DATA_TYPE  | IS_NULLABLE | COLUMN_DEFAULT | CHARACTER_MAXIMUM_LENGTH | NUMERIC_PRECISION | NUMERIC_SCALE |
+-------------+------------+-------------+----------------+--------------------------+-------------------+---------------+
| text        | text       | YES         | NULL           |                    65535 |              NULL |          NULL |
| tinytext    | tinytext   | YES         | NULL           |                      255 |              NULL |          NULL |
| mediumtext  | mediumtext | YES         | NULL           |                 16777215 |              NULL |          NULL |
| longtext    | longtext   | YES         | NULL           |               4294967295 |              NULL |          NULL |
+-------------+------------+-------------+----------------+--------------------------+-------------------+---------------+
4 rows in set (0.00 sec)
```

In function `toType()`, when case is `charLen.Valid` we return `schema.Type{Name: dataType, Mods: []int64{charLen.Int64}}`.  Thus, for `text` type, mods will be allocated, and hence I have changed to`maxExpectedMods(1)` in `mysql/toddl.go` file.

File: mysql/infoschema.go (hunk 3 above)
Comment (nevinheintze) [reply mysql/infoschema.go:27]: Many thanks for the detailed response.

What's your take on RawBytes vs interface{}?

My understanding was that interface{} would return values from the underlying driver without conversion.
But RawBytes effectively forces conversion to a string format. Which means that if the value is an int64, then we do int64 -> string -> int64.

But I 100% agree with you about the redundant data conversion code.

Should we just do both postgres and mysql via string? Any thoughts on the additional processing/conversion cost?

File: mysql/infoschema.go [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+// getConstraints returns a list of primary keys and by-column map of
+// other constraints.  Note: we need to preserve ordinal order of
+// columns in primary key constraints.
+func getConstraints(conv *internal.Conv, db *sql.DB, table schemaAndName) ([]string, map[string][]string, error) {
+	q := `SELECT k.COLUMN_NAME, t.CONSTRAINT_TYPE
+              FROM INFORMATION_SCHEMA.TABLE_CONSTRAINTS AS t
+                INNER JOIN INFORMATION_SCHEMA.KEY_COLUMN_USAGE AS k
+                  ON t.CONSTRAINT_NAME = k.CONSTRAINT_NAME AND t.CONSTRAINT_SCHEMA = k.CONSTRAINT_SCHEMA AND t.TABLE_NAME=k.TABLE_NAME
+              WHERE k.TABLE_SCHEMA = ? AND k.TABLE_NAME = ? ORDER BY k.ordinal_position;`
```
Comment (nevinheintze) [inline mysql/infoschema.go:277]: Why use '?' instead of $1, $2 for the parameters?

File: mysql/infoschema.go [hunk 6]
```diff
@@ -0,0 +1,364 @@
...
+		return nil, nil, err
+	}
+	defer rows.Close()
+	var primaryKeys []string
+	var col, constraint string
+	m := make(map[string][]string)
+	for rows.Next() {
+		err := rows.Scan(&col, &constraint)
+		if err != nil {
+			fmt.Printf("Can't scan: %v\n", err)
```
Comment (nevinheintze) [inline mysql/infoschema.go:289]: This is going to potentially spam stdout with lots of errors.
Maybe better as:

conv.Unexpected(fmt.Sprintf("Can't scan: %v\n", err))

This is also an issue in postgres/infoschema.go.

Same comment in other uses of fmt.Printf in this file.

File: mysql/infoschema.go (hunk 5 above)
Comment (Dharmil-searce) [reply mysql/infoschema.go:277]: MySQL does not support bind parameters `$1,$2..` like postgres. It throws an error when such bind parameters are used:
```
panic: couldn't get schema for table testdb.Persons: Error 1054: Unknown column '$1' in 'where clause'
```
Thus, used `?` instead of `$1,$2..`.

File: mysql/infoschema.go (hunk 6 above)
Comment (Dharmil-searce) [reply mysql/infoschema.go:289]: Agreed. Initially, I had put it as unexpected only but then changed it to fmt.Printf() after comparing the code with postgres/infoschema.go.
Sure, I will update `fmt.Printf()` usage with `conv.Unexpected(fmt.Sprintf())` in both the packages.

File: mysql/infoschema.go [hunk 7]
```diff
@@ -0,0 +1,364 @@
...
+	case "YES":
+		return false
+	case "NO":
+		return true
+	}
+	conv.Unexpected(fmt.Sprintf("isNullable column has unknown value: %s", isNullable))
+	return false
+}
+
+// buildVals contructs interface{} value containers to scan row
```
Comment (nevinheintze) [inline mysql/infoschema.go:332]: You are constructing []RawBytes containers.

How about rephrasing:

// buildVals constructs []sql.RawBytes value containers to scan row

File: mysql/infoschema.go [hunk 8]
```diff
@@ -0,0 +1,364 @@
...
+// results into.  Returns both the underlying containers (as a slice)
+// as well as an interface{} of pointers to containers to pass to
+// rows.Scan.
+func buildVals(n int) (v []sql.RawBytes, iv []interface{}) {
+	// Make a slice for the values
```
Comment (nevinheintze) [inline mysql/infoschema.go:337]: Suggest deleting this comment -- its not adding much, given the function-level comment.

File: mysql/infoschema_test.go [hunk 9]
```diff
@@ -0,0 +1,304 @@
...
+		"bs": []internal.SchemaIssue{internal.DefaultValue},
+		"f4": []internal.SchemaIssue{internal.Widened},
+		"i4": []internal.SchemaIssue{internal.Widened},
+		"i2": []internal.SchemaIssue{internal.Widened},
+		"si": []internal.SchemaIssue{internal.Widened, internal.DefaultValue},
+		"ts": []internal.SchemaIssue{internal.Datetime},
+	}
+	assert.Equal(t, expectedIssues, conv.Issues["test.test"])
+	assert.Equal(t, int64(0), conv.Unexpecteds())
+}
```
Comment (nevinheintze) [inline mysql/infoschema_test.go:144]: Should there be a TestProcessSqlData similar to the one in postgres/infoschema_test.go?

File: mysql/infoschema_test.go [hunk 10]
```diff
@@ -0,0 +1,304 @@
...
+
+// TODO : Remove this test (Not needed)
```
Comment (nevinheintze) [inline mysql/infoschema_test.go:146]: Yes, agreed. I think we can remove this since it is redundant.

File: mysql/infoschema.go (hunk 8 above)
Comment (Dharmil-searce) [reply mysql/infoschema.go]: Done.

File: mysql/infoschema.go [hunk 11]
```diff
@@ -0,0 +1,364 @@
...
+	// TODO: refactor to use the set of tables computed by
+	// ProcessInfoSchema instead of computing them again.
+	tables, err := getTables(db)
+	if err != nil {
+		conv.Unexpected(fmt.Sprintf("Couldn't get list of table: %s", err))
+		return
+	}
+	for _, t := range tables {
+		srcTable := buildTableName(t.schema, t.name)
+		srcSchema, ok1 := conv.SrcSchema[srcTable]
```
[the remark this answers is not in the document]
Comment (Dharmil-searce) [reply mysql/infoschema.go]: Done.

File: mysql/infoschema.go (hunk 7 above)
Comment (Dharmil-searce) [reply mysql/infoschema.go]: Done.

File: mysql/infoschema.go (hunk 6 above)
Comment (Dharmil-searce) [reply mysql/infoschema.go]: Updated in both `mysql/infoschema.go` and `postgres/infoschema.go`.

File: mysql/toddl.go (hunk 4 above)
Comment (nevinheintze) [reply mysql/toddl.go:139]: Thanks for all of the details and the example.

So this doesn't show up in mysql dump parsing, just in the info schema tables?

In effect we're relaxing the error checking for mysql dump generated schemas, but I don't see a huge issue in doing that.

File: mysql/infoschema.go (hunk 5 above)
Comment (nevinheintze) [reply mysql/infoschema.go:277]: Ah right. Got it. That's super annoying.
I thought mysql did support $1, $2 etc.
Maybe its an issue with the golang database/sql driver?

File: mysql/toddl.go (hunk 4 above)
Comment (Dharmil-searce) [reply mysql/toddl.go:139]: Correct, In` mysqldump` create statements will have `text` and `blob` datatype without any mods. It's also mentioned in pingcap documentation under data types section ([LINK](https://docs.pingcap.com/tidb/dev/sql-statement-create-table)).
Even if mysqldump has a modifier for `text` & `blob` datatype then also pingcap parser will ignore it and return the parsed statement without any modifier ( I have tested the same for both datatypes). Hence, we are not relaxing error checking for mysqldump. 
One work around could be to explicitly put a condition when datatype is `text` or `blob` and return `schema.Type{Name: dataType}` in mysql/infoschema.go  and then we do not need to change `maxExpectedMods(0)` but I do not see much benefit of doing that and also it would not be a good code design.

Thus, I think we can move forward with this change of error checks.

File: mysql/infoschema.go (hunk 5 above)
Comment (Dharmil-searce) [reply mysql/infoschema.go:277]: Yeah, maybe it's not supported by golang driver.

File: mysql/infoschema_test.go [hunk 12]
```diff
@@ -0,0 +1,304 @@
...
+			ColDefs:  map[string]ddl.ColumnDef{col: ddl.ColumnDef{Name: col, T: tc.spType, IsArray: tc.isArray}}}
+		conv.TimezoneOffset = "+05:30"
+		_, ac, av, err := ConvertData(conv, tableName, cols, srcSchema, tableName, cols, spSchema, []string{tc.in})
+		assert.Equal(t, cols, ac)
+		assert.Equal(t, []interface{}{tc.e}, av)
+		assert.Nil(t, err)
+	}
+}
+
+func TestConvertSqlRow_MultiCol(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (Dharmil-searce) [reply mysql/infoschema_test.go]: Done. Renamed.

File: mysql/infoschema_test.go (hunk 9 above)
Comment (Dharmil-searce) [reply mysql/infoschema_test.go:143]: In `ProcessSQLData` function of `mysql/infoschema.go` we are calling `ProcessDataRow` of `mysql/data.go` unlike in postgres where `ConvertSQLRow` is being called. Hence, adding `TestProcessSqlData` is not adding any benefit because test cases for `ProcessDataRow` are covered in `mysql/data_test.go` file.

Still, as per your suggestion I have added `TestProcessSQLData`.

**Comment by Dharmil-searce** [comment]
> I'm happy with the PR.
> Please merge when all comments are resolved.

@nevinheintze Only those with write access to the repo can merge PRs. I do not have that access so request you to merge the PR.

> Should we just do both postgres and mysql via string? Any thoughts on the additional processing/conversion cost?

We can discuss this on the group.

File: mysql/infoschema.go (hunk 3 above)
Comment (Dharmil-searce) [reply mysql/infoschema.go:27]: Main benefit to use RawBytes is to avoid redundancy in code and reuse data conversion functions. Other benefit is that we can avoid complex switch conditions in `convScalar` and `convArray`.
It would definitely be easier and in sync to use RawBytes in postgres as well. [This](github.com/lib/pq) library is currently used in postgres. Need to check if it supports RawBytes. Also, in the documentation of current library another library([LINK](https://github.com/jackc/pgx)) is recommended to be used. In addition to that,array structure is a bit complex in postgres compared to mysql need to check that also.

We can discuss more about additional processing/conversion cost on google chat group.
